teps/0120-concurrency-controls.md

```diff
@@ -0,0 +1,102 @@
+last-updated: '2022-08-19'
+authors:
+- '@vdemeester'
+- '@williamlfish'
+- '@lbernick'
+---
+
+# TEP-0120: Concurrency Controls
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Use Cases](#use-cases)
+  - [Requirements](#requirements)
+  - [Future Work](#future-work)
+- [Proposal](#proposal)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Allow users to configure desired behavior for managing concurrent PipelineRuns.
+
+## Motivation
+
+Allow users to avoid wasting resources on redundant work, to prevent concurrent runs of
+non-idempotent operations, and to have more control over concurrent load on a cluster.
+
+### Use Cases
+
+Avoiding redundant work
+- A developer pushes a pull request and quickly notices and fixes a typo. They would like to have the first CI run automatically canceled and replaced by the second.
+
+Controlling non-idempotent operations
+- An organization uses a Pipeline for deployment, and wants to make sure that only the most recent changes are applied to their cluster.
+If a new deployment PipelineRun starts while a previous one is still running, they would like the previous one to be canceled.
+  - The organization might want only one deployment PipelineRun per cluster, per namespace, per environment (prod/staging),
+  per repo, or per user ([example](https://github.com/tektoncd/pipeline/issues/2828#issuecomment-646150534)).
+- An integration test communicates with a stateful external service (like a database), and a developer wants to ensure
+that integration testing TaskRuns within their CI PipelineRun don’t run concurrently with other integration testing TaskRuns
+(based on [this comment](https://github.com/tektoncd/pipeline/issues/2828#issuecomment-647747330)).
+- An organization would like to ensure Terraform operations run serially (based on [this comment](https://github.com/tektoncd/experimental/issues/699#issuecomment-951606279))
+  - This user would like to be able to cancel queued runs that are pending, rather than cancelling running runs. 
+
+Controlling load on a cluster
+- An organization has multiple teams working on a mobile application with a limited number of test devices.
+They want to limit the number of concurrent CI runs per team, to prevent one team from using all the available devices
+and crowding out CI runs from other teams.
+- A Pipeline performs multiple parallelizable things with different concurrency caps, as described in [this comment](https://github.com/tektoncd/pipeline/issues/2591#issuecomment-626778025).
+- Allow users to cap the number of matrixed TaskRuns (alpha) that can run at a given time.
+  - Currently, we have the feature flag “default-maximum-matrix-fan-out”, which restricts the total number of TaskRuns
+  that can be created from one Pipeline Task. However, we would like to support capping the number of matrixed TaskRuns
+  that can run concurrently, instead of statically capping the number of matrixed TaskRuns that can be created at all.
```

Comment:
This use case is referring to `TaskRuns` only - can it be expanded to `TaskRuns`, `Runs` and `PipelineRuns`? Or is there an underlying assumption that `TaskRuns` is used to refer to all the run types? After rename `Runs` to `CustomRuns` (TEP-0114), we could agree to use `Runs` to refer to all the run types

Reply:
I agree that it would be good to use "runs" to refer to taskRuns, pipelineRuns, and customRuns. I've updated the wording here to make it clear that this use case is for all types of matrixed runs.